### DISCUSSION TITLE
How to retain background knowledge associations in the original data when using MWEM synthesizer

### DISCUSSION THREAD

**Discussion by Mistin1885** [opendp/opendp#630] [SmartNoise] [answered]
I would like to ask if there are any synthesizer parameters that preserve the attribute associations in the original data.

For example, using a dataset containing `age` and `ever_married`, where `ever_married` has only Yes or No values. It can be observed in the raw data that when `age` is less than 18, the value of `ever_married` must be No.

However, when I looked at the data generated by the synthesizer, I found that this rule does not exist in the synthesized data. I tried to use the 'splits' parameter to set their dependencies, but it doesn't seem to work.

Thanks

**Comment by lurosenb** [comment, marked as answer]
Hi Mistin1885,

## Issue with hard constraints based on sensitive data
In order to preserve differential privacy, we cannot **place hard constraints** on the output of our synthesizers based on our input data! This is why such constraints are not built into any of the smartnoise synthesizer offereings. 

Let me give you a bit of intuition based on the ε-DP definition:

ε-DP - Given a function `f` that outputs a summary of some dataset, and given two datasets `D1` and `D2` that differ on a single element (representing the data of one individual), `f` is said to have ε-differential privacy if, for all possible outputs `S` of `f` and for all possible datasets `D1` and `D2`:

`P[f(D1) = S] / P[f(D2) = S] ≤ e^ε`

Here, `P[f(D) = S]` is the probability of `f` producing output `S` when its input is `D`.

Maybe you see the issue - the case where we constrain the function `f(D)` to produce a synthetic dataset `S` containing rows where `age` is less than 18 and the value of `ever_married` is `No` based on input data `D` implies that the probability of producing an `S` where `age` is less than 18 and the value of `ever_married` is `Yes` is 0! This means that we end up with an infinite or undefined value of ε, which is no good. However, not all hope is lost

## Solution with public data
Luckily, we can run our differentially private distribution, and then apply post-hoc constraints based on publicly available data. This does not violate our privacy definition, as this has no dependency on the input data and thus is not potentially leaking any information about it! 

A good example of when this is applicable is in learning an "Age" feature in a dataset. We know that no human has ever lived past 122 - this is public knowledge. So if we see rows where "Age>122", we can throw them out. Because we can sample arbitrarily from our privately learned distribution, post-processing consistency constraints based on public knowledge is a good way to ensure that you get realistic values from your private data. If you have public data that suggests that no one under the `Age` of 18 is `ever_married`, then this would be a reasonable post-processing step to employ. But it should never be part of the synthesizers themselves, based on the input data, otherwise we violate our privacy guarantee : )
